x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java

```diff
@@ -493,14 +506,118 @@ public Expression visitNullLiteral(NullLiteralContext ctx) {
+                        + "use the string notation instead", trailing);
+            }
+            value = of(interval.valueNumeric, negative, leading);
+            valueAsText = interval.valueNumeric.getText();
+        } else {
+            value = visitIntervalValue(interval.valuePattern, negative, intervalType);
+            valueAsText = interval.valuePattern.getText();
+        }
+
+        String name = "INTERVAL " + valueAsText + " " + leading.name() + (trailing != null ? " TO " + trailing.name() : "");
+
+        Interval<?> timeInterval = value instanceof Period ? new IntervalYearMonth((Period) value,
+                intervalType) : new IntervalDayTime((Duration) value, intervalType);
+
+        return new Literal(source(ctx), name, timeInterval, intervalType);
+    }
+
+    private TemporalAmount of(NumberContext valueNumeric, boolean negative, TimeUnit unit) {
+        // expect numbers for now
+        Literal value = (Literal) visit(valueNumeric);
+        Number numeric = (Number) value.fold();
+
+        if (Math.rint(numeric.doubleValue()) != numeric.longValue()) {
+            throw new ParsingException(source(valueNumeric), "Fractional values are not supported for intervals");
+        }
+
+        return IntervalUtils.of(source(valueNumeric), numeric.longValue(), unit);
+    }
+
+    private TemporalAmount visitIntervalValue(StringContext valuePattern, boolean negative, DataType intervalType) {
+        String valueString = string(valuePattern);
+        Location loc = source(valuePattern);
+        TemporalAmount interval = IntervalUtils.parseInterval(loc, valueString, intervalType);
+        if (negative) {
+            interval = IntervalUtils.negate(interval);
+        }
+        return interval;
+    }
+
+    @Override
+    public IntervalUtils.TimeUnit visitIntervalField(IntervalFieldContext ctx) {
+        if (ctx == null) {
+            return null;
+        }
+
+        switch (ctx.getChild(TerminalNode.class, 0).getSymbol().getType()) {
+            case SqlBaseParser.YEAR:
+            case SqlBaseParser.YEARS:
+                return IntervalUtils.TimeUnit.YEAR;
+            case SqlBaseParser.MONTH:
+            case SqlBaseParser.MONTHS:
+                return IntervalUtils.TimeUnit.MONTH;
+            case SqlBaseParser.DAY:
+            case SqlBaseParser.DAYS:
+                return IntervalUtils.TimeUnit.DAY;
+            case SqlBaseParser.HOUR:
+            case SqlBaseParser.HOURS:
+                return IntervalUtils.TimeUnit.HOUR;
+            case SqlBaseParser.MINUTE:
+            case SqlBaseParser.MINUTES:
+                return IntervalUtils.TimeUnit.MINUTE;
+            case SqlBaseParser.SECOND:
+            case SqlBaseParser.SECONDS:
+                return IntervalUtils.TimeUnit.SECOND;
+        }
+
+        throw new IllegalArgumentException("Unsupported interval field: " + ctx.getText());
```

Comment:
You could also put it as `default` of the switch block.

Reply:
True however I wasn't sure whether we would do any extra processing of the interval field hence why I left it as a method catch-all.